components/encoder/collectors_test.go

```diff
@@ -0,0 +1,71 @@
+package encoder
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	clk "github.com/benbjohnson/clock"
+	"github.com/edaniels/golog"
+	pb "go.viam.com/api/component/encoder/v1"
+	"go.viam.com/test"
+	"go.viam.com/utils/protoutils"
+
+	"go.viam.com/rdk/data"
+	"go.viam.com/rdk/resource"
+	tu "go.viam.com/rdk/testutils"
+)
+
+const componentName = "encoder"
+
+func TestEncoderCollector(t *testing.T) {
+	mockClock := clk.NewMock()
+	buf := tu.MockBuffer{}
+	params := data.CollectorParams{
+		ComponentName: componentName,
+		Interval:      time.Second,
+		Logger:        golog.NewTestLogger(t),
+		Target:        &buf,
+		Clock:         mockClock,
+	}
+
+	enc := newEncoder(componentName)
+	col, err := newTicksCountCollector(enc, params)
+	test.That(t, err, test.ShouldBeNil)
+
+	defer col.Close()
+	col.Collect()
+	mockClock.Add(1 * time.Second)
+
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, len(buf.Writes), test.ShouldEqual, 1)
+	test.That(t, buf.Writes[0].GetStruct().AsMap(), test.ShouldResemble, toProtoMap(pb.GetPositionResponse{
+		Value:        1.0,
+		PositionType: pb.PositionType_POSITION_TYPE_TICKS_COUNT,
+	}))
+}
+
+type fakeEncoder struct {
```

Comment:
Maybe a dumb question because I haven't used it much - but I see we have all these very simple mocks for each of these. Is this something that that existing `inject` (or whatever it's called) package can already provide?

Reply:
Importing the inject package creates a circular import loop as the inject package imports the component.

Reply:
We discussed with @dmhilly about this offline as a clean way to get around importing the injected arm and having the circular dependency. The only thing I worry about is if the Go method signature changes, especially the internal Go struct return types — but if it did, then the collector code itself shouldn't compile in the first place.

Another way to included the `inject` package would be to make this test file in a different package and export the `newXCollector` constructors. We already exported `RegisterCollector` in the `data` package, and I don't think there's much harm — if another package tries to reregister on the same method, we correctly return an error (or more specifically, panic).

Reply:
> Another way to included the inject package would be to make this test file in a different package and export the newXCollector constructors. We already exported RegisterCollector in the data package, and I don't think there's much harm — if another package tries to reregister on the same method, we correctly return an error (or more specifically, panic).

To clarify - the suggestion is to move collectors.go and collectors_test.go to a new package (or just the latter)? I think moving both would make sense, and (at least at first glance) I don't see anything that would break

Reply:
Just the latter, since I believe moving both would create a circular dependency where the collector brings in the component but the component needs to depend on the collector.

Reply:
Chatted offline and Vijay's making these changes — thank you! This'll keep the test code robust to any method signature changes on the component.